crates/wasi-common/src/fdentry.rs

```diff
@@ -1,35 +1,335 @@
 use crate::sys::dev_null;
 use crate::sys::fdentry_impl::{determine_type_and_access_rights, OsFile};
+use crate::hostcalls_impl::Dirent;
 use crate::{wasi, Error, Result};
 use std::path::PathBuf;
-use std::{fs, io};
+use std::{fs, io, fmt};
+use std::path::Path;
+use std::cell::RefCell;
+use std::io::SeekFrom;
+use filetime::FileTime;
+
+use std::sync::Arc;
+use std::collections::HashMap;
+use std::collections::hash_map::Entry;
+
+pub trait VirtualFile {
+    // methods that virtual files need to have to uh, work right?
+    fn fdstat_get(&self) -> wasi::__wasi_fdflags_t {
+        0
+    }
+
+    fn try_clone(&self) -> io::Result<Box<dyn VirtualFile>>;
+
+    fn readlinkat(&self, path: &Path) -> Result<String>;
+
+    fn openat(&self, path: &Path, read: bool, write: bool, oflags: u16, fs_flags: u16) -> Result<Box<dyn VirtualFile>>;
+
+    fn datasync(&self) -> Result<()> {
+        Err(Error::EINVAL)
+    }
+
+    fn sync(&self) -> Result<()> {
+        Ok(())
+    }
+
+    fn create_directory(&self, path: &Path) -> Result<()> {
+        Err(Error::EACCES)
+    }
+
+    fn readdir(&self, cookie: wasi::__wasi_dircookie_t) -> Result<Box<dyn Iterator<Item = Result<Dirent>>>> {
+        Err(Error::EBADF)
+    }
+
+    fn write_vectored(&mut self, iovs: &[io::IoSlice]) -> Result<usize> {
+        Err(Error::EBADF)
+    }
+
+    fn pread(&self, buf: &mut [u8], offset: u64) -> Result<usize> {
+        Err(Error::EBADF)
+    }
+
+    fn pwrite(&self, buf: &mut [u8], offset: u64) -> Result<usize> {
+        Err(Error::EBADF)
+    }
+
+    fn seek(&mut self, offset: SeekFrom) -> Result<u64> {
+        Err(Error::EBADF)
+    }
+
+    fn advise(&self, advice: wasi::__wasi_advice_t, offset: wasi::__wasi_filesize_t, len: wasi::__wasi_filesize_t) -> Result<()> {
+        Err(Error::EBADF)
+    }
+
+    fn allocate(&self, offset: wasi::__wasi_filesize_t, len: wasi::__wasi_filesize_t) -> Result<()> {
+        Err(Error::EBADF)
+    }
+
+    fn filestat_get(&self) -> Result<wasi::__wasi_filestat_t> {
+        Err(Error::EBADF)
+    }
+
+    fn filestat_set_times(&self, atim: Option<FileTime>, mtim: Option<FileTime>) -> Result<()> {
+        Err(Error::EBADF)
+    }
+
+    fn filestat_set_size(&self, st_size: wasi::__wasi_filesize_t) -> Result<()> {
+        Err(Error::EBADF)
+    }
+
+    fn fdstat_set_flags(&self, fdflags: wasi::__wasi_fdflags_t) -> Result<()> {
+        Err(Error::EBADF)
+    }
+
+    fn read_vectored(&mut self, iovs: &mut [io::IoSliceMut]) -> Result<usize> {
+        Err(Error::EBADF)
+    }
+
+    fn get_file_type(&self) -> wasi::__wasi_filetype_t;
+
+    fn get_rights_base(&self) -> wasi::__wasi_rights_t;
+
+    fn get_rights_inheriting(&self) -> wasi::__wasi_rights_t;
+}
+
+pub struct InMemoryFile {
```

Comment:
`InMemoryFile` is a byproduct of working as a result of adjusting `lucet-wasi`. It seems like this would be general enough that it would be of interest to `wasi-common` as well, but obligating `wasi-common` to have a functional virtual filesystem seems excessive. I just don't know where else I'd put this at the moment.